### PR TITLE
add the register_mxid_from_3pid setting (untested)

### PR DESCRIPTION
**Pull request by ara4n**
Adds a setting to prohibit users from picking their own mxid, but instead derive it from the 3PID that they provide during signup.

I need to test this and then seek review from @dbkr probably - opening the PR in order to link to it from elsewhere.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/register.py [hunk 1]
```diff
@@ -376,6 +373,69 @@ def on_POST(self, request):
...
+                                    )
+                                else:
+                                    desired_username += "1"
+                            else:
+                                # something else went wrong.
+                                break
+
+                    # XXX: a nasty heuristic to turn an email address into
+                    # a displayname, as part of register_mxid_from_3pid
+                    parts = address.replace('.', ' ').split('@')
```
Comment (dbkr) [inline synapse/rest/client/v2_alpha/register.py]: You're doing this for msisdns too - is this going to fail for these (presumably there just won't be a parts[1])

File: synapse/rest/client/v2_alpha/register.py [hunk 2]
```diff
@@ -376,6 +373,69 @@ def on_POST(self, request):
...
+                                assigned_user_id=registered_user_id,
+                            )
+                            # if we got this far we passed the check.
+                            break
+                        except SynapseError as e:
+                            if e.errcode == Codes.USER_IN_USE:
+                                m = re.match(r'^(.*)(\d+)$', desired_username)
+                                if m:
+                                    desired_username = m.group(1) + str(
+                                        int(m.group(2)) + 1
```
Comment (dbkr) [inline synapse/rest/client/v2_alpha/register.py]: Adding a '1' on the end could be a bit confusing for msisdns

**Comment by ara4n** [comment]
@dbkr hopefully ^ fixes that.

**Comment by ara4n** [comment]
well, that was a mess. going to merge now, but some retrospective review from @dbkr would be appreciated, especially as it touches on the profile replication stuff.

**Pull request merged**